src/persist-client/src/operators/shard_source.rs

```diff
@@ -298,117 +222,277 @@ where
+        let (subscription, as_of) = match futures::future::select(token_is_dropped, create_subscribe).await {
+            Either::Left(_) => {
+                // the token dropped before we finished creating our Subscribe.
+                // we can return immediately, as we could not have emitted any
+                // parts to fetch.
+                cap_set.downgrade(&[]);
+                return;
+            }
+            Either::Right(((subscription, as_of), _)) => {
+                (subscription, as_of)
+            }
+        };
+
+        let mut subscription = subscription.unwrap_or_else(|e| {
+            panic!(
+                "{}: {} cannot serve requested as_of {:?}: {:?}",
+                name_owned, shard_id, as_of, e
+            )
+        });
+
+        let mut batch_parts = vec![];
+        let mut lease_returner = subscription.lease_returner().clone();
+        let (_keep_subscribe_alive_tx, keep_subscribe_alive_rx) = tokio::sync::oneshot::channel::<()>();
+        let subscription_stream = async_stream::stream! {
+            // Eagerly yield the initial as_of. This makes sure that the output
+            // frontier of the `persist_source` closely tracks the `upper` frontier
+            // of the persist shard. It might be that the snapshot for `as_of` is
+            // not initially available yet, but this makes sure we already downgrade
+            // to it.
+            //
+            // Downstream consumers might rely on close frontier tracking for making
+            // progress. For example, the `persist_sink` needs to know the
+            // up-to-date upper of the output shard to make progress because it
+            // will only write out new data once it knows that earlier writes went
+            // through, including the initial downgrade of the shard upper to the
+            // `as_of`.
+            yield ListenEvent::Progress(as_of.clone());
+
+            let mut done = false;
+            while !done {
+                for event in subscription.next().await {
+                    if let ListenEvent::Progress(ref progress) = event {
+                        // If `until.less_equal(progress)`, it means that all subsequent batches will
+                        // contain only times greater or equal to `until`, which means they can be
+                        // dropped in their entirety. The current batch must be emitted, but we can
+                        // stop afterwards.
+                        if PartialOrder::less_equal(&until, progress) {
+                            done = true;
                         }
+                    }
+                    yield event;
+                }
+            }
+
+            // intentionally keep this stream from dropping until the operator
+            // is dropped. this ensures our Subscribe handle stays alive for as
+            // long as is needed to finish fetching all of its parts.
+            let _ = keep_subscribe_alive_rx.await;
```

Comment:
This relies on `_keep_subscribe_alive_tx` not leaving scope/being dropped?

Reply:
How do we make sure that doesn't happen? (I'm probably not getting some of the rust details ... 🙈 )

Reply:
it relies on `_keep_subscribe_alive_tx` only being dropped when the whole logic fn returns, which should only happen when our feedback edge reaches the empty frontier

Reply:
is that true with NLL (non-lexical lifetimes)?

Reply:
this might be interacting differently with the async closure, so, ... 🤷‍♂️😅

Reply:
no idea about NLL, but `_keep_subscribe_alive_tx` is in the scope of the rest of the closure, not within the async_stream, so I'd think it wouldn't drop until the closure returns/drops? if it doesn't, this would really break my understanding of how any RAII guards work